libgadget/forcetree.c

```diff
@@ -715,50 +727,52 @@ static int
 force_update_node_recursive(int no, int sib, int level, const ForceTree * tree, const int HybridNuGrav)
 {
     /*Last value of tails is the return value of this function*/
-    int j, suns[8], tails[8];
+    int j, suns[NCHILD], tails[NCHILD];
 
-    /*Count how many internal children we have,
-     *so we can keep track of how many tasks we started*/
-    int chldcnt=0;
     /* this "backup" is necessary because the nextnode
      * entry will overwrite one element (union!) */
-    for(j = 0; j < 8; j++) {
-        suns[j] = tree->Nodes[no].u.suns[j];
-        if(suns[j] >= tree->firstnode && suns[j] < tree->lastnode)
-            chldcnt++;
+    int noccupied = tree->Nodes[no].u.s.noccupied;
+    for(j = 0; j < NCHILD; j++) {
+        suns[j] = tree->Nodes[no].u.s.suns[j];
     }
 
+    int limit = noccupied;
+    int childcnt = 0;
+    if(noccupied >= 1<<16) {
+        limit = 8;
+        /* Remove empty nodes from the tree*/
+        for(j=0; j < 8; j++)
+            if(suns[j] < tree->lastnode) {
+                if(tree->Nodes[suns[j]].u.s.noccupied == 0)
+                    suns[j] = -1;
+                else
+                    childcnt++;
+            }
+    }
     /*First do the children*/
-    for(j = 0; j < 8; j++)
+    for(j = 0; j < limit; j++)
     {
         int p = suns[j];
-        /*Empty slot*/
-        if(p < 0)
+        if(suns[j] < 0)
             continue;
         /* For particles and pseudo particles we have nothing to update; */
         /* But the new tail is the last particle in the linked list. */
         if(p < tree->firstnode || p >= tree->lastnode) {
```

Comment:
Why is it no longer necessasry to find the correct tail?

Reply:
We are still finding the correct tail: the logic is to ensure the tails are not confused by the presence of empty subnodes

Reply:
The loop is not needed anymore because we don't have a linked list of sub particles

Reply:
Are you confident the old loop was only looping through particles inside the same node? get_next_node doesn't guarantee ending at a node boundary -- it would go through any partially created traversal link-list..

Reply:
That is true, but the old code (somewhat obscurely) set nextnode = -1 by hand on every particle added to the tree, so the lists were always of length 0 unless the node was too small. The code is:

        /*If the node is too small we prepend the particle to a short linked list.*/
        force_set_next_node(p_toplace, too_small ? p_child : -1, &tb);
So before it was indeed only looping through particles inside the same node.